File: frontend/src/api/asset.ts
import http from "./http";

export interface Asset {
  id: string;
  user_id: string;
  name: string;
  description: string;
  category: "model_3d" | "model_live2d" | "motion" | "texture";
  format: string;
  file_key: string;
  file_size: number;
  thumbnail_key: string;
  metadata: Record<string, any>;
  tags: string[];
  is_public: boolean;
  download_count: number;
  status: "processing" | "ready" | "failed";
  created_at: string;
  updated_at: string;
}

export interface CompleteUploadParams {
  file_key: string;
  name: string;
  description?: string;
  category: string;
  format: string;
  file_size: number;
  is_public?: boolean;
  tags?: string[];
}

export interface UpdateAssetParams {
  name?: string;
  description?: string;
  is_public?: boolean;
  tags?: string[];
}

export const assetApi = {
  getPresignedUpload: (filename: string, category: string, fileSize: number) =>
    http.post<{ data: { upload_url: string; file_key: string } }>(
      "/assets/upload/presigned",
      { filename, category, file_size: fileSize },
    ),
  completeUpload: (data: CompleteUploadParams) =>
    http.post<{ data: Asset }>("/assets/upload/complete", data),
  list: (page = 1, pageSize = 20, category?: string, format?: string) =>
    http.get("/assets", { params: { page, page_size: pageSize, category, format } }),
  listPublic: (page = 1, pageSize = 20, category?: string, search?: string) =>
    http.get("/assets/public", { params: { page, page_size: pageSize, category, search } }),
  get: (id: string) => http.get(`/assets/${id}`),
  update: (id: string, data: UpdateAssetParams) => http.put(`/assets/${id}`, data),
  delete: (id: string) => http.delete(`/assets/${id}`),
  getDownloadURL: (id: string) =>
    http.get<{ data: { url: string } }>(`/assets/${id}/download`),
  getThumbnailURL: (id: string) =>
    http.get<{ data: { url: string } }>(`/assets/${id}/thumbnail`),
};

File: backend/internal/model/asset.go
package model

import (
	"time"

	"github.com/lib/pq"
)

// Asset represents an uploaded model or motion file entity.
type Asset struct {
	ID            string         `gorm:"type:char(26);primaryKey" json:"id"`
	UserID        string         `gorm:"type:char(26);index;not null;column:user_id" json:"user_id"`
	Name          string         `gorm:"type:varchar(255);not null" json:"name"`
	Description   string         `gorm:"type:text" json:"description"`
	Category      string         `gorm:"type:varchar(32);not null;index" json:"category"`
	Format        string         `gorm:"type:varchar(16);not null;index" json:"format"`
	FileKey       string         `gorm:"type:varchar(512);not null;column:file_key" json:"file_key"`
	FileSize      int64          `gorm:"not null;column:file_size" json:"file_size"`
	ThumbnailKey  string         `gorm:"type:varchar(512);column:thumbnail_key" json:"thumbnail_key"`
	Metadata      string         `gorm:"type:jsonb;default:'{}'" json:"metadata"`
	Tags          pq.StringArray `gorm:"type:text[];default:'{}'" json:"tags"`
	IsPublic      bool           `gorm:"default:false;column:is_public;index" json:"is_public"`
	DownloadCount int            `gorm:"default:0;column:download_count" json:"download_count"`
	Status        string         `gorm:"type:varchar(16);default:processing;not null" json:"status"`
	CreatedAt     time.Time      `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt     time.Time      `gorm:"autoUpdateTime" json:"updated_at"`

	User User `gorm:"foreignKey:UserID" json:"-"`
}

// TableName specifies the database table name for the Asset model.
func (Asset) TableName() string {
	return "assets"
}
